src/Raven.Server/Documents/Indexes/Static/JavaScriptIndex.cs

```diff
@@ -0,0 +1,298 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Dynamic;
+using System.Linq;
+using System.Text;
+using Esprima;
+using Esprima.Ast;
+using Jint;
+using Jint.Native;
+using Jint.Native.Array;
+using Jint.Native.Function;
+using Lucene.Net.Util;
+using Raven.Client.Documents.Indexes;
+using Raven.Server.Documents.Patch;
+
+namespace Raven.Server.Documents.Indexes.Static
+{
+    public class JavaScriptIndex : StaticIndexBase
+    {
+        private readonly Engine _jint;
+
+        public JavaScriptIndex(IndexDefinition definition)
+        {            
+            _definitions = definition;
+            _jint = new Engine();
+            _jint.Execute(Code);
+            foreach (var map in definition.Maps)
+            {
+                _jint.Execute(map);
+                //ExtractFields(map);
+            }
+            var definitions = _jint.GetValue("globalDefinition").AsObject();
+            var maps = definitions.GetProperty("maps").Value.AsArray();
+            var mapLength = maps.GetLength();
+            _collectionFunctions = new Dictionary<string, List<List<JsValue>>>();
+            for (int i = 0; i < mapLength; i++)
+            {
+                var map = maps.Get(i.ToString()).AsObject();
+                var mapCollection = map.Get("Collection").AsString();
+                if (_collectionFunctions.TryGetValue(mapCollection, out var list) == false)
+                {
+                    list = new List<List<JsValue>>();
+                    _collectionFunctions.Add(mapCollection, list);
+                }
+                var chainList = new List<JsValue>();
+                var mapChain = map.Get("Chain").AsArray();
+
+                var chainLength = mapChain.GetLength();
+                for (int j = 0; j < chainLength; j++)
+                {
+                    var method = mapChain.Get(j.ToString());
+                    chainList.Add(method);
+                }
+                list.Add(chainList);
+            }
+            var fields = new HashSet<string>();
+            foreach (var (key, val) in _collectionFunctions)
+            {
+                Maps.Add(key, val.Select(x => (IndexingFunc)(enm => JintMapFuncWrapper(x, enm))).ToList());
```

Comment:
Need to be explicit that you are passing the _functions_ here.

Reply:
I'll probably modify this and create an object and pass its member function to make this more clear.